Improvements to RepRap5d Driver

bjj opened the pull request:
My github branch has a few things going on so you might want to pick and choose.  I just did a trial merge (after the revert described below) and the only issues are changelog (entries can be combined) and the sf39 changes (since sf39 is gone).  You can see my merged tree at bjj:merge.  The windows launcher is not working for me in that tree, however.

Extensive improvements to the RepRap5d driver.  These have been tested by me with FiveD, Klimentkip (related to Tonokip, Klimentkip) and Teacup.  Others have also reported success.  The only outstanding issues I'm aware of appear to be serial port problems.  The few areas that touch common code have been regression tested against a Thing-o-Matic.  You need to revert 89aff4cf from makerbot:master before applying my changes.

A few skeinforge enhancements (although my branch is pre-sf40).  Things like putting outline.py and reversal.py back in sf39, making dimension.py work with reversal.

Rolled back the linux .so's for RXTX serial to 2.1.7 to match the Java jar version.

A few misc bugfixes.

cibomahto commented:
Hi Ben,

Any chance you could clean this up so it merges with the current master? I'll promise not to break anything new for a short time. I've got a reasonably large change (home_positioning branch) that I'm planning to merge soon, and it might be a good idea to get your work in first.

Cheers,
Matt

bjj commented:
My trial merge is at:  https://github.com/bjj/ReplicatorG/commit/d73a2a1ded249ebbc8279806fee7e71742377cf2  .  I just did a test of merging that commit onto your master and it is automatic.  Note that there's a later commit on that branch (bjj:merge) where I fix ReplicatorG.exe and if you try to merge that it conflicts because of the deletion in makerbot:master.

The only parts of my tree that were lost were some sf39 changes which could apply to sf40.  I can do that separately if they're appropriate and submit another pull request.

koenkooi commented:
Since the list of patches is not that long, you could try 'git cherry-pick' on the ones you want, that usually avoids the merge problems. Or do 'git rebase -i origin/master' and delete the skeinforge commits.

koenkooi commented:
Be carefull with rebase -i, commits will really be gone, so do it on a test branch!

bjj commented:
On Tue, Apr 19, 2011 at 11:28:10AM -0700, koenkooi wrote:

> Be carefull with rebase -i, commits will really be gone, so do it on a test branch!

Commits are never lost in git (at least until you gc in a few weeks).
But if you rebase -i on a pushed branch and re-push you will confuse
others who have pulled.

## 

Ben Jackson AD7GD
ben@ben.com
http://www.ben.com/

bjj commented:
I added another commit (e8f998a31 in bjj:merge, also on master) to fix an ordering problem in sf-35 extrusion.py:  dimensino needs to be nearly last in order to translate all M motor commands into E.

koenkooi commented:
What actually curious about is why most pullrequests to repG try to fix multiple problems, 5d, skeinforge, gen6 and rxtx in this one, instead of opening 4 different ones.
If the repG people don't mind, more power to them!

bjj commented:
On Mon, Apr 25, 2011 at 06:17:51AM -0700, koenkooi wrote:

> What actually curious about is why most pullrequests to repG try to fix multiple problems, 5d, skeinforge, gen6 and rxtx in this one, instead of opening 4 different ones.
> If the repG people don't mind, more power to them!

Does it really matter that much how a pull request breaks down?  The
changes themselves are broken into something like 30 distinct checkins
with individual descriptions.  I expect them to take or leave them
individually as they see fit.

## 

Ben Jackson AD7GD
ben@ben.com
http://www.ben.com/

cibomahto commented:
@koenkooi: Part of the issue is that the code isn't well separated, so it becomes difficult in many instances to fix one thing without touching multiple parts of the code

@bjj: It is important to separate out issues, because we have to verify that each piece works before pulling it in. If the pull request is broken into several smaller, atomic chunks, then it is much safer and straightforward to pull in.

bjj commented:
Actually if you want to test RepG after each piece then it's less work to accept a single large patch and test once.  Certainly it's less work for me to present one monolithic patch that I've tested extensively than to try to verify that many independent changes can work in any combination you choose.  I commit small changes in what I think is a rational order to give other people more flexibility in how they use the patches, but some fixes were developed together and only split up at commit time.

If someone with commit authority wants to make a specific request about how to present these changes, now is the time.  The only request I got is "make it merge with master" so I did the merge, tested the merge, and then made fixes to bugs introduced on master unrelated to my code so I could test the result.  If @koenkooi is willing to accept pull requests of a certain scope and size, then be explicit about the parameters rather than just complaining "this is big".   I see this pull request as the opening bid in a negotiation:  If you have reasonable requests about how patches should be submitted I will try to accommodate them.  If you have _un_reasonable requests (in my opinion, of course) then I will continue to enjoy my modified RepG and not trouble you with further patches.

koenkooi commented:
@bjj: I don't have anything to do with the repG development team :) I was just wondering why a lot of pull requests for repG mix things. For the projects I work on pull requests tend to contain only fixes to one area. 
It wasn't meant as a criticism to anyone, just trying to clear up what the repG people would like. I got the (possibly wrong) impression that @cibomahto is reluctant to pull in your 5D fixes because there are other changes in between like the skeinforge and rxtx ones.

@cibomahto: is there any documentation that shows how you want your pull requests? It looks like repG is getting popular enough to need one :)

koenkooi commented:
@bjj: and IMO a pull request with 30 tidy commits is just as good as one single patch. Actually, if it isn't a collection of commit/revert pairs it's a lot better than one single patch. But that's a personal preference, I know a lot of people who disagree with me.

cibomahto commented:
@bjj @koenkooi: Lets move this discussion over to the mailing list. A guideline is a good idea, but its an orthogonal issue from this pull request.

bjj committed:
machines/reprap.xml: Add Klimentkip and Teacup entries.

bjj committed:
Allow reconcilePosition() to return null to getCurrentPosition()

Indicates no position change.  Fixes infinite loop in RepRap5D caused by
recursing into getCurrentPosition() trying to avoid a position change.
